src/3d/qgsframegraph.cpp

```diff
@@ -45,8 +45,300 @@ typedef Qt3DCore::QGeometry Qt3DQGeometry;
+#include <Qt3DRender/QSeamlessCubemap>
 #include <Qt3DRender/QNoDepthMask>
 #include <Qt3DRender/QBlendEquationArguments>
+#include <Qt3DExtras/QTextureMaterial>
+#include <Qt3DRender/QAbstractTexture>
+
+namespace
+{
+  QString formatIdName( quint64 id, const QString &name )
+  {
+    QString n = name.isEmpty() ? QLatin1String( "<no_name>" ) : name;
+    return QLatin1String( "{%1/%2}" ).arg( QString::number( id ) ).arg( n );
+  }
+
+  QString formatIdName( const Qt3DRender::QAbstractTexture *texture )
+  {
+    QString n = texture->objectName().isEmpty() ? QLatin1String( "<no_name>" ) : texture->objectName();
+    return QLatin1String( "{%1[%2]/%3" )
+           .arg( QString::number( texture->id().id() )
+                 , QString( QMetaEnum::fromType<Qt3DRender::QAbstractTexture::TextureFormat>().valueToKey( texture->format() ) )
+                 , n );
+  }
+
+  QString formatNode( const Qt3DCore::QNode *n )
+  {
+    QString res = QLatin1String( "(%1%2)" )
+                  .arg( QLatin1String( n->metaObject()->className() ) )
+                  .arg( formatIdName( n->id().id(), n->objectName() ) );
+    if ( !n->isEnabled() )
+      res += QLatin1String( " [D]" );
+    return res;
+  }
+
+  QString formatList( const QStringList &lst )
+  {
+    return QString( QLatin1String( "[ %1 ]" ) ).arg( lst.join( QLatin1String( ", " ) ) );
+  }
+
+  QString formatLongList( const QStringList &lst, int level )
+  {
+    QString out = formatList( lst );
+    if ( out.size() < 200 )
+      return out;
+
+    out = QString( QLatin1String( "[\n" ) );
+    for ( QString l : lst )
+    {
+      l = QString( "-> %1\n" ).arg( l );
+      out += l.rightJustified( l.length() + ( 1 + level ) * 2, ' ' );
+    }
+    QString end( QLatin1String( "]" ) );
+    return out + end.rightJustified( end.length() + ( 1 + level ) * 2, ' ' );
+  }
+
+  QString formatField( const QString &name, const QString &value )
+  {
+    if ( value == "<no_value>" )
+      return QString( QLatin1String( "(%1)" ) ).arg( name );
+    return QString( QLatin1String( "(%1:%2)" ) ).arg( name, value );
+  }
+
+  QString dumpSGEntity( const Qt3DCore::QEntity *n, int level )
+  {
+    auto extractTextureParam = []( const QVector<Qt3DRender::QParameter *> &params, QStringList & fl )
+    {
+      for ( const auto *param : params )
+      {
+        if ( strstr( param->value().typeName(), "QAbstractTexture*" ) != nullptr )
+        {
+          const Qt3DRender::QAbstractTexture *tex = param->value().value<Qt3DRender::QAbstractTexture *>();
+          fl += formatField( param->name(), formatIdName( tex ) );
+        }
+      }
+    };
+
+
+    QString res = formatNode( n );
+    const auto &components = n->components();
+    if ( ! components.isEmpty() )
+    {
+      QStringList componentNames;
+      for ( const auto &c : components )
+      {
+        QString res = formatNode( c );
+        QStringList fl;
+        if ( qobject_cast<const Qt3DExtras::QTextureMaterial *>( c ) )
```

Comment:
could do something like:
```
if ( auto c_cast  = qobject_cast<...>(c) )
{
...
}
```
to avoid the double cast (not a big deal but more readable I think)

Reply:
done!